types/src/database_sql.rs

```diff
@@ -226,6 +225,83 @@ impl Database {
         }))
     }
 
+    pub async fn upsert_collection(
+        &self,
+        title: String,
+        wordpress_menu_id: i64,
+        slug: String) -> Result<String>{
+        query_file!(
+            "queries/upsert_collection.sql",
+            slug,
+            title,
+            wordpress_menu_id)
+            .execute(&self.client)
+            .await?;
+        Ok(slug)
+    }
+
+    pub async fn insert_all_chapters(
+        &self,
+        chapters: Vec<Chapter>,
+        slug: String) -> Result<String>{
+
+        let mut tx = self.client.begin().await?;
+
+        // Delete all chapter data before re-inserting
+        query_file!(
+            "queries/delete_chapters.sql")
+        .execute(&mut tx)
```

Comment:
Jesus, no, don't delete _all_ chapters. This function needs to be scoped to a specific collection, so that you only delete the chapters under `cwkw` then insert all the `cwkw` chapters.

Reply:
Right, I kind forgot that we would eventually have multiple edited collections so that's why I had done that-- what would the best way to go about doing this be? I was thinking since each chapter path would have the collection slug included in it, I could just delete every chapter where the collection slug forms part of the chapter path, but I'm not sure if there's an easier/simpler/better way to do that.

Reply:
Yeah that's exactly what I'd do, delete chapters where the very first element in the path is exactly the collection slug. The [ltree docs](https://www.postgresql.org/docs/current/ltree.html) show some examples of these queries. I _think_ the best way to do this exact query (rather than just checking if `cwkw` is any of the ancestors or just somewhere in the path), is: `chapter.path ~ format('%s.*', $1)`

That checks that `chapter.path` matches the query "any path that starts with the string $1 and has anything after that".

Reply:
I bet you could come up with another way of writing that query that doesn't use `format` too.

Reply:
Hmm, when I tried the expression you suggested I got errors since it interpreted format('%s.*', $1) as text-- maybe I missed it but I couldn't find a way to convert from text to lquery within a query in the documentation.

What I ended up doing was using PgLQuery::from_str outside of the query itself, after concatenating the slug with ".*". This way also doesn't use format, but I'm not sure if there's a more intuitive/better way to write it.

Reply:
Ohhh that should work!